### ISSUE TITLE
EmpiricalMarginal.mean can return nan; make test problem [bug]

### ISSUE THREAD

**Issue by adam-coogan** [uber/pyro#1548]
I'm starting to use pyro for a (physics) research project and have been testing different posterior inference methods on very simple models. I've found that `EmpiricalMarginal.mean` returns `nan` when `_log_weights` is small. I'm on the dev branch and can provide an example if it would be helpful.

The problem stems from this helper method:

https://github.com/uber/pyro/blob/5643a749b7a7c4a654ae41d7b9212dd67411542f/pyro/distributions/empirical.py#L132-L137

When `max_val.exp()` is very small, `_weighted_mean()` returns zero, causing `mean()` to return a tensor of `nan`s. However, the `max_val.exp()` factors will always cancel in the numerator and denominator when computing the mean:

https://github.com/uber/pyro/blob/5643a749b7a7c4a654ae41d7b9212dd67411542f/pyro/distributions/empirical.py#L155

The same `_weighted_mean()` function appears in two other places:

        pyro/examples/rsa/search_inference.py
        pyro/tutorial/source/search_inference.py

It's simple to fix this and I would submit a pull request, but `make test` is failing on my machine with the following error:

```
[gw4] [ 74%] FAILED tests/ops/test_stats.py::test_autocorrelation 
...
    def autocorrelation(input, dim=0):
        """
        Computes the autocorrelation of samples at dimension `dim`.
    
        Reference: https://en.wikipedia.org/wiki/Autocorrelation#Efficient_computation
    
        :param torch.Tensor input: the input tensor.
        :param int dim: the dimension to calculate autocorrelation.
        :returns torch.Tensor: autocorrelation of `input`.
        """
        if (not input.is_cuda) and (not torch.backends.mkl.is_available()):
>           raise NotImplementedError("For CPU tensor, this method is only supported "
                                      "with MKL installed.")
E           NotImplementedError: For CPU tensor, this method is only supported with MKL installed.
```

I have MKL installed. Here's my environment info:
-OS: macOS Mojave 10.14.1 (18B75)
-`python -c 'import torch; print(torch.__version__)'` returns `0.4.0`
-`python -c 'import pyro; print pyro.__version__'` returns `0.2.1`
Does someone know how to fix this? I'd really like to contribute to pyro, even if it's just a small thing like this!

**Comment by fehiepsi** [comment]
Nice observation! For mkl, you can just ignore it I guess. You don’t have to redo all the tests. If you want to deal with that issue, I suggest to try calling ‘torch.backends.mkl.is_available()’ to see if you have installed torch with mkl backend or not.

**Comment by neerajprad** [comment]
Thanks for reporting - It will be great if you could submit your PR fixing this. Don't worry about this particular test, but it will be nice to add your failing example to `test_empirical.py` and assert that it passes now.. 

@fehiepsi - I think we should just skip or xfail if we don't have mkl available, I am not sure how it is passing on travis.

**Comment by neerajprad** [comment]
Okay, I see why this wasn't failing on travis. That is using the wheels that I built with MKL and not the default wheels. @fritzo has addressed this issue in #1550.

**Comment by adam-coogan** [comment]
Closed by https://github.com/uber/pyro/pull/1558
